src/serial.rs

```diff
@@ -0,0 +1,538 @@
+                pub fn listen(&mut self, _event: Event) {
+                    unimplemented!();
+                }
+
+                /// Stop listening for an interrupt event
+                pub fn unlisten(&mut self, _event: Event) {
+                    unimplemented!();
+                }
+
+                /// Return true if the receiver is idle
+                pub fn is_rx_idle(& self) -> bool {
+                    self.uart.status.read().st_urx_out().is_rx_idle()
+                }
+
+                /// Return true if the transmitter is idle
+                pub fn is_tx_idle(& self) -> bool {
+                    self.uart.status.read().st_utx_out().is_tx_idle()
+                }
+
+                pub fn split(self) -> (Tx<$UARTX>, Rx<$UARTX>) {
+                    (
+
+                        Tx {
+                            _uart: PhantomData,
+                            _apb_lock: if let None=self.apb_lock { None } else { Some(self.clock_control.lock_apb_frequency()) }
+                        },
+                        Rx {
+                            _uart: PhantomData,
+                            _apb_lock: if let None=self.apb_lock { None } else { Some(self.clock_control.lock_apb_frequency()) }
+                        },
+                    )
+                }
+
+                pub fn release(self) -> ($UARTX, PINS) {
+                    (self.uart, self.pins)
+                }
+
+            }
+
+            impl<'a, PINS> serial::Read<u8> for Serial<$UARTX, PINS> {
+                type Error = Infallible;
+
+                fn read(&mut self) -> nb::Result<u8, Self::Error> {
+                    let mut rx: Rx<$UARTX> = Rx {
+                        _uart: PhantomData,
+                        _apb_lock: if let None=self.apb_lock { None } else { Some(self.clock_control.lock_apb_frequency()) }
+                    };
+                    rx.read()
+                }
+            }
+
+
+            impl  Rx<$UARTX> {
+                pub fn count(& self) -> u8 {
+                    unsafe {
+                            (*$UARTX::ptr()).status.read().rxfifo_cnt().bits()
+                        }
+                }
+
+                pub fn is_idle(& self) -> bool {
+                    unsafe {
+                            (*$UARTX::ptr()).status.read().st_urx_out().is_rx_idle()
+                        }
+                }
+            }
+
+            impl  Tx<$UARTX> {
+                pub fn count(& self) -> u8 {
+                    unsafe {
+                            (*$UARTX::ptr()).status.read().txfifo_cnt().bits()
+                        }
+                }
+
+                pub fn is_idle(& self) -> bool {
+                    unsafe {
+                            (*$UARTX::ptr()).status.read().st_utx_out().is_tx_idle()
+                        }
+                }
+            }
+
+            impl serial::Read<u8> for Rx<$UARTX> {
+                type Error = Infallible;
```

Comment:
Is this really an infallible operation? Surely there are bus errors etc, to look for?

Reply:
There could be parity errors etc, but these are reported asynchronously (and optionally via interrupts). So I don't think it is possible to report them form this function.

Reply:
I agree, but I think they can also be read from the status register itself. I think `Register 13.3: UART_INT_ST_REG` on page 354 in the reference manual seems to have what were looking for?

Reply:
Yes but which byte in the FIFO do they belong to?

We can of course just report, when the api user happens to read, but the meaning will then be: an error occured on this byte or one of the future bytes in the FIFO.

Reply:
Hmm, I see your point. I'll take a look at how the idf handles errors and report back.

Reply:
Will take a further look later on as well, but propose to delay this to a next pull request anyway. Is that ok?